### DISCUSSION TITLE
Ebitengine 2.5 stability

### DISCUSSION THREAD

**Discussion by divVerent** [hajimehoshi/ebiten#2569] [General]
I right now am very tempted to release the next build of my game using v2.5.0-alpha.12, but am worried that it may not be "finished enough" yet, or that I may be causing more harm than gain by doing so (e.g. because APIs may still be subject to change, or there being known major issues that 2.4 does not have).

E.g. I have noticed this: my game AAAAXY runs well in both 2.4 and 2.5 now, but especially on a lowend Android phone (Moto G7 Play), 2.5 has much more constant performance:

![ebi24](https://user-images.githubusercontent.com/251568/217712071-611842b8-263f-4ed6-991d-e52d60a17559.png)

In 2.4, there are many frames that take substantially longer than average.

![ebi25](https://user-images.githubusercontent.com/251568/217712079-00dd4ab3-0142-4b96-a2ae-14eb472c453d.png)

In 2.5, there are only few "lag frames" left.

This difference is rather noticeable, so I'd really like to upgrade - but would like to first clear up if 2.5-alpha in its current shape is "releasable" if one accepts the things that are not done yet and that some APIs for new features may still change, or if it better should be avoided for stability reasons and I should rather stick with 2.4 until 2.5 at least reaches rc status (which I assume requires https://github.com/hajimehoshi/ebiten/milestone/82 to be fulfilled).

Thanks!

**Comment by hajimehoshi** [comment]
Thank you for taking the performance! What do the axes mean in the graphs?

During alpha versions, there is no guarantee that we don't add breaking changes to the APIs. So if you worried about the stability of the API, I recommmend sticking to v2.4.

Note that there is no further plan to add API changes to v2.5 until RC releases so far. However, this is not 100%.

**Reply by hajimehoshi** [reply]
@divVerent Could you move your reply here? :-)

**Reply by divVerent** [reply]
Oh sorry, I should have labelled.

X axis is frame number of benchmark.dem in the game - Y axis is
milliseconds per frame. All at default settings - performance can be
improved by reducing settings, so no worries about this graph indicating
just around 15 fps on either side of the comparison.

The interesting part is how more constant timing got. This is rather
quickly noticeable when playing the game, at any settings.

API stability is not a big worry for me as I notice issues when compiling.
Primarily wondering about stability of the code - are there major issues
still expected to be lurking in there, or known "mysterious crashes"? I see
one potential issue regarding WebGL performance and SubImage texture render
targets that I still need to try out, but nothing serious in the open
tickets otherwise.

(Sorry, I had forgotten that replying by email is broken on github)

**Reply by hajimehoshi** [reply]
> Primarily wondering about stability of the code - are there major issues
> still expected to be lurking in there, or known "mysterious crashes"? 

No, there are not such issues. There are still known crashes, but they are pretty rare and not easy to fix so I might leave them in v2.5.

> I see
> one potential issue regarding WebGL performance and SubImage texture render
> targets that I still need to try out, but nothing serious in the open
> tickets otherwise.

Yes there is such this issue in WebGL, but this is not a regression from v2.4. I think it seems impossible to fix unless we adopt WebGPU.

**Reply by divVerent** [reply]
Thanks; did an easy workaround for subimage-as-render-target then (there's only one case where I did that, and that call was _usually_ redundant so just not calling it if the size already matches anyway was simple enough) and am now strongly considering switching to 2.5-alpha, provided it's causing no issues on any platforms I test on.

Will however also look into how I can make it easy to switch back to 2.4 if needed, despite having code changes in there for DrawFinalScreen. Build tags, I guess ;)

**Reply by hajimehoshi** [reply]
Yeah, I think it is a good timing to switch to v2.5 now, as v2.5 is already quite matured. I'm considering to release RCs this month btw.

**Reply by hajimehoshi** [reply]
v2.5.0 has already been released :-)
